### PR TITLE
Added vertices for cheated (non-neutrino) particles.

### PR DESCRIPTION
**Pull request by StevenGreen1**
This fix allows Track and Shower objects to be made from cheated test beam particles.  Currently, cheated cosmic rays and test beam particles cannot be used to make tracks or showers inside LArSoft as there is no vertex made when cheating.  

This change consists of a configurable option to add a start vertex to all PFOs made in the LArCheatingPfoCreation algorithm.  If this flag is set to true in the cosmic ray pass of the LArCheatingPfoCreation algorithm, start vertices are added to all cosmic rays and test beam particles, which then allows them to be made into tracks and showers.

There are other options available to us on how we could proceed with this fix, such as only adding start vertices to test beam particles (as the start vertex for a cosmic ray could be ambiguous), which we should discuss before merging this request.  This fix has been tested inside LArSoft and does work for single electron showers.

### CODE REVIEW

File: larpandoracontent/LArCheating/CheatingPfoCreationAlgorithm.cc [hunk 1]
```diff
@@ -129,6 +130,27 @@ void CheatingPfoCreationAlgorithm::CreatePfos(const MCParticleToClusterListMap &
 
             const ParticleFlowObject *pPfo(nullptr);
             PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::ParticleFlowObject::Create(*this, pfoParameters, pPfo));
+
+            if (m_addVertices)
+            {
+                const VertexList *pVertexList = NULL; std::string vertexListName;
```
Comment (PandoraPFA) [inline larpandoracontent/LArCheating/CheatingPfoCreationAlgorithm.cc]: `nullptr` now preferred (has a pointer type), and via copy constructor.

File: larpandoracontent/LArCheating/CheatingPfoCreationAlgorithm.cc [hunk 2]
```diff
@@ -129,6 +130,27 @@ void CheatingPfoCreationAlgorithm::CreatePfos(const MCParticleToClusterListMap &
...
+                PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::CreateTemporaryListAndSetCurrent(*this, pVertexList, vertexListName));
+
+                PandoraContentApi::Vertex::Parameters parameters;
+                parameters.m_position = CartesianVector(pMCParticle->GetVertex().GetX(), pMCParticle->GetVertex().GetY(), pMCParticle->GetVertex().GetZ());
```
Comment (PandoraPFA) [inline larpandoracontent/LArCheating/CheatingPfoCreationAlgorithm.cc]: Can just assign `pMCParticle->GetVertex()` directly.

File: larpandoracontent/LArCheating/CheatingPfoCreationAlgorithm.cc [hunk 3]
```diff
@@ -129,6 +130,27 @@ void CheatingPfoCreationAlgorithm::CreatePfos(const MCParticleToClusterListMap &
...
+                parameters.m_vertexLabel = VERTEX_START;
+                parameters.m_vertexType = VERTEX_3D;
+
+                const Vertex *pVertex(NULL);
```
Comment (PandoraPFA) [inline larpandoracontent/LArCheating/CheatingPfoCreationAlgorithm.cc]: `nullptr`

File: larpandoracontent/LArCheating/CheatingPfoCreationAlgorithm.cc [hunk 4]
```diff
@@ -129,6 +130,27 @@ void CheatingPfoCreationAlgorithm::CreatePfos(const MCParticleToClusterListMap &
 
             const ParticleFlowObject *pPfo(nullptr);
             PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::ParticleFlowObject::Create(*this, pfoParameters, pPfo));
+
+            if (m_addVertices)
+            {
+                const VertexList *pVertexList = NULL; std::string vertexListName;
+                PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::CreateTemporaryListAndSetCurrent(*this, pVertexList, vertexListName));
```
Comment (PandoraPFA) [inline larpandoracontent/LArCheating/CheatingPfoCreationAlgorithm.cc]: Probably better to just create the temporary list once, as for the Pfos, above. Then fill it as we build Pfos and save it at the end.

File: larpandoracontent/LArCheating/CheatingPfoCreationAlgorithm.cc [hunk 5]
```diff
@@ -129,6 +130,27 @@ void CheatingPfoCreationAlgorithm::CreatePfos(const MCParticleToClusterListMap &
...
+
+                PandoraContentApi::Vertex::Parameters parameters;
+                parameters.m_position = CartesianVector(pMCParticle->GetVertex().GetX(), pMCParticle->GetVertex().GetY(), pMCParticle->GetVertex().GetZ());
+                parameters.m_vertexLabel = VERTEX_START;
+                parameters.m_vertexType = VERTEX_3D;
+
+                const Vertex *pVertex(NULL);
+                PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::Vertex::Create(*this, parameters, pVertex));
+
+                if (!pVertexList->empty())
```
Comment (PandoraPFA) [inline larpandoracontent/LArCheating/CheatingPfoCreationAlgorithm.cc]: Suggest saving all vertices in one go, at end, just as for Pfos.

Can still add Vertex to Pfo at this point, but an exception will be raised at the end of the algorithm if you don't save a vertex that has been added to a Pfo. Pandora will attempt to delete the temporary Vertex, and find it is in use!

File: larpandoracontent/LArCheating/CheatingPfoCreationAlgorithm.cc [hunk 6]
```diff
@@ -129,6 +130,27 @@ void CheatingPfoCreationAlgorithm::CreatePfos(const MCParticleToClusterListMap &
...
+                {
+                    PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::SaveList<Vertex>(*this, m_vertexListName));
+                    PANDORA_THROW_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::ReplaceCurrentList<Vertex>(*this, m_vertexListName));
```
Comment (PandoraPFA) [inline larpandoracontent/LArCheating/CheatingPfoCreationAlgorithm.cc]: Might want to add option as to whether you really want to replace the current list. Basically, and as for most of the above, suggest that you just follow what is already done for the Pfos.

**Comment by StevenGreen1** [comment]
Thanks for the comments, I've implemented the changes, but would appreciate another quick check just to make sure I understood what you meant with the vertex list.  

I tested this out on an event and it seemed to work as expected (vertices produced for everything with an option of whether to replace the current list or not).  Also as we discussed I have changed the default behaviour to be to create the vertices for the PFO, but to not replace the current vertex list.  

I believe this means anyone wanting to build Tracks and Showers in LArSoft using cheated PFParticles won't have to alter PandoraSettings_Cheat.xml too, which is nice.

**Comment by PandoraPFA** [comment]
All good - merged via commandline.

**Pull request closed without merging**